components/engine_skiplist/src/engine.rs

```diff
@@ -0,0 +1,511 @@
+                    if key > u {
+                        self.last_kv = None;
+                        return Ok(false);
+                    }
+                }
+                if let Some(e) = self.engine.lower_bound(Bound::Included(key)) {
+                    if check_in_range(
+                        e.key(),
+                        self.upper_bound.as_ref(),
+                        self.lower_bound.as_ref(),
+                    ) {
+                        Some((e.key().to_vec(), e.value().to_vec()))
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            }
+        };
+        Ok(self.last_kv.is_some())
+    }
+    fn seek_for_prev(&mut self, key: SeekKey) -> Result<bool> {
+        let _timer = SKIPLIST_ACTION_HISTOGRAM_VEC
+            .with_label_values(&["seek_for_prev"])
+            .start_coarse_timer();
+
+        match key {
+            SeekKey::Start | SeekKey::End => self.seek(key),
+            SeekKey::Key(key) => {
+                if let Some(l) = self.lower_bound.as_deref() {
+                    if key < l {
+                        return Ok(false);
+                    }
+                }
+                if let Some(u) = self.upper_bound.as_deref() {
+                    if key > u {
+                        self.last_kv = None;
+                        return self.seek(SeekKey::End);
+                    }
+                }
+                self.last_kv = if let Some(e) = self.engine.upper_bound(Bound::Excluded(key)) {
+                    if check_in_range(
+                        e.key(),
+                        self.upper_bound.as_ref(),
+                        self.lower_bound.as_ref(),
+                    ) {
+                        Some((e.key().to_vec(), e.value().to_vec()))
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                };
+                Ok(self.last_kv.is_some())
+            }
+        }
+    }
+
+    fn prev(&mut self) -> Result<bool> {
+        let _timer = SKIPLIST_ACTION_HISTOGRAM_VEC
+            .with_label_values(&["prev"])
+            .start_coarse_timer();
+        if self.last_kv.is_none() {
+            return Ok(false);
+        }
+        let (last_key, _) = self.last_kv.as_ref().unwrap();
+        self.last_kv = if let Some(e) = self
+            .engine
+            .upper_bound(Bound::Excluded(last_key.as_slice()))
+        {
+            if check_in_range(
+                e.key(),
+                self.upper_bound.as_ref(),
+                self.lower_bound.as_ref(),
+            ) {
+                Some((e.key().to_vec(), e.value().to_vec()))
+            } else {
+                None
+            }
+        } else {
+            None
+        };
+        Ok(self.last_kv.is_some())
+    }
+    fn next(&mut self) -> Result<bool> {
```

Comment:
Why seek rather than move pointer of skiplist?

Reply:
If this SkipListMap does not support 'Iterator', why do not we implement a new one? Or we can write a rust-wrapper for `rocksdb::Skiplist`

Reply:
The original version implements the iterator in this way, but it is extremely slow because keeping the point of a node makes much cost for waiting spinlock.